File: python/algo/202606/algo0843.py
from typing import List

# 力扣内置类，提交时不用写
class Master:
    def guess(self, word: str) -> int:
        pass

class Solution:
    def findSecretWord(self, words: List[str], master: 'Master') -> None:
        # 极小化极大算法
        candidate = words.copy()

        def similar(a: str, b: str) -> int:
            cnt = 0
            for i in range(6):
                if a[i] == b[i]:
                    cnt += 1
            return cnt

        for _ in range(30):
            n = len(candidate)
            select_idx = 0
            min_max = n

            # 选出最优猜测词（ minimax ）
            for i in range(n):
                freq = [0] * 6
                for j in range(n):
                    if i == j:
                        continue
                    s = similar(candidate[i], candidate[j])
                    freq[s] += 1
                current_max = max(freq)
                if current_max < min_max:
                    min_max = current_max
                    select_idx = i

            # 猜测
            guess_word = candidate[select_idx]
            res = master.guess(guess_word)
            if res == 6:
                return

            # 筛选候选集
            next_candidate = []
            for j in range(n):
                if select_idx == j:
                    continue
                if similar(candidate[select_idx], candidate[j]) == res:
                    next_candidate.append(candidate[j])

            candidate = next_candidate

File: cplus/leetcode/202606/algo0843.cpp
#include <bits/stdc++.h>
using namespace std;

// 贪心 - 七、交互题
/**
 * // This is the Master's API interface.
 * // You should not implement it, or speculate about its implementation
 */
class Master {
   public:
      int guess(string word);
};

class Solution {
public:
    void findSecretWord(vector<string>& words, Master& master) {
        // 参考题解，极小化极大值
        vector<string> candicate(words);

        auto similiar = [](string& a, string& b) -> int {
            int cnt = 0;
            for (int i=0; i<6; i++) if (a[i] == b[i]) cnt ++;
            return cnt;
        };

        for (int _=0; _<30; _++) {
            int n = candicate.size(), inx = 0, minmax = n;
            for (int i=0; i<n; i++) {                
                vector<int> sim(n), freq(6);
                for (int j=0; j<n; j++) {
                    if (i == j) continue;
                    sim[j] = similiar(candicate[i], candicate[j]);
                    freq[sim[j]] ++;
                }
                auto mx = ranges::max(freq);
                if (mx < minmax) minmax = mx, inx = i;
            }
            cout << candicate[inx] << endl;
            auto gus = master.guess(candicate[inx]);
            if (gus == 6) return;
            vector<string> nxt;
            for (int j=0; j<n; j++)
                if (inx!=j && similiar(candicate[inx], candicate[j]) == gus)
                    nxt.push_back(candicate[j]);
            
            candicate = move(nxt);  
        }
    }
};